File: full-stack-assignment-app/aws-backend/assignment_functions/get_assignments.py
import json
import boto3
import os
from decimal import Decimal

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')

def lambda_handler(event, context):
    """
    Lambda function to retrieve assignments from DynamoDB
    """
    
    # CORS headers
    cors_headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization',
        'Content-Type': 'application/json'
    }
    
    try:
        # Handle CORS preflight requests
        if event.get('httpMethod') == 'OPTIONS':
            return {
                'statusCode': 200,
                'headers': cors_headers,
                'body': json.dumps({'message': 'CORS preflight handled'})
            }
        
        # Only allow GET requests
        if event.get('httpMethod') != 'GET':
            return {
                'statusCode': 405,
                'headers': cors_headers,
                'body': json.dumps({'error': 'Method not allowed. Use GET.'})
            }
        
        # Get DynamoDB table
        table_name = os.environ.get('DYNAMODB_TABLE_NAME', 'AssignmentsTable')
        table = dynamodb.Table(table_name)
        
        # Check if specific assignment ID is requested
        assignment_id = None
        if 'pathParameters' in event and event['pathParameters']:
            assignment_id = event['pathParameters'].get('assignmentId')
        
        if assignment_id:
            # Get specific assignment
            response = table.get_item(Key={'assignmentId': assignment_id})
            
            if 'Item' not in response:
                return {
                    'statusCode': 404,
                    'headers': cors_headers,
                    'body': json.dumps({'error': 'Assignment not found'})
                }
            
            assignment = convert_decimal_to_float(response['Item'])
            
            return {
                'statusCode': 200,
                'headers': cors_headers,
                'body': json.dumps({
                    'success': True,
                    'assignment': assignment
                })
            }
        
        else:
            # Get all assignments (scan - use with caution in production)
            response = table.scan()
            assignments = [convert_decimal_to_float(item) for item in response['Items']]
            
            # Sort by creation date (newest first)
            assignments.sort(key=lambda x: x.get('createdAt', ''), reverse=True)
            
            return {
                'statusCode': 200,
                'headers': cors_headers,
                'body': json.dumps({
                    'success': True,
                    'assignments': assignments,
                    'count': len(assignments)
                })
            }
        
    except Exception as error:
        # Log the error
        print(f"Error retrieving assignments: {str(error)}")
        print(f"Event: {json.dumps(event)}")
        
        # Return error response
        return {
            'statusCode': 500,
            'headers': cors_headers,
            'body': json.dumps({
                'success': False,
                'error': 'Internal server error',
                'message': str(error)
            })
        }

def convert_decimal_to_float(obj):
    """
    Convert Decimal values to float for JSON serialization
    """
    if isinstance(obj, list):
        return [convert_decimal_to_float(item) for item in obj]
    elif isinstance(obj, dict):
        return {key: convert_decimal_to_float(value) for key, value in obj.items()}
    elif isinstance(obj, Decimal):
        return float(obj)
    else:
        return obj

File: full-stack-assignment-app/aws-backend/assignment_functions/get_assignments.mjs
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, ScanCommand, GetCommand } from "@aws-sdk/lib-dynamodb";

// Initialize DynamoDB client
const client = new DynamoDBClient({});
const dynamodb = DynamoDBDocumentClient.from(client);

export const handler = async (event) => {
    // CORS headers
    const corsHeaders = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-Amz-Date, X-Api-Key, X-Amz-Security-Token',
        'Content-Type': 'application/json'
    };
    
    try {
        console.log('Event received:', JSON.stringify(event, null, 2));
        
        // Handle CORS preflight OPTIONS request
        if (event.httpMethod === 'OPTIONS') {
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: JSON.stringify({ message: 'CORS preflight successful' })
            };
        }
        
        // Only allow GET requests (OPTIONS handled above)
        if (event.httpMethod !== 'GET') {
            return {
                statusCode: 405,
                headers: corsHeaders,
                body: JSON.stringify({ error: 'Method not allowed. Use GET.' })
            };
        }
        
        // Get DynamoDB table name from environment variable
        const tableName = process.env.DYNAMODB_TABLE_NAME || 'AssignmentsTable';
        console.log('Using table:', tableName);
        
        // Check if specific assignment ID is requested
        let assignmentId = null;
        if (event.pathParameters && event.pathParameters.assignmentId) {
            assignmentId = event.pathParameters.assignmentId;
        }
        
        if (assignmentId) {
            // Get specific assignment
            console.log('Getting specific assignment:', assignmentId);
            const command = new GetCommand({
                TableName: tableName,
                Key: { assignmentId: assignmentId }
            });
            
            const response = await dynamodb.send(command);
            
            if (!response.Item) {
                return {
                    statusCode: 404,
                    headers: corsHeaders,
                    body: JSON.stringify({ error: 'Assignment not found' })
                };
            }
            
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: true,
                    assignment: response.Item
                })
            };
        } else {
            // Get all assignments - IMPROVED FILTERING
            console.log('Getting all assignments from table:', tableName);
            const command = new ScanCommand({
                TableName: tableName
            });
            
            const response = await dynamodb.send(command);
            const allItems = response.Items || [];
            
            console.log('Found total items before filtering:', allItems.length);
            
            // Filter to get only actual assignments (exclude student responses)
            const assignments = allItems.filter(item => {
                // Student responses have responseId in format: student#ID#assignment#ID
                if (item.responseId) {
                    console.log('Filtering out student response:', item.responseId);
                    return false;
                }
                
                // Student responses have type = 'assignment-response'
                if (item.type === 'assignment-response') {
                    console.log('Filtering out response by type:', item.type);
                    return false;
                }
                
                // Items with assignmentId and title are likely assignments
                if (item.assignmentId && item.title) {
                    console.log('Keeping assignment:', item.assignmentId, item.title);
                    return true;
                }
                
                // Log items that don't match any criteria for debugging
                console.log('Unknown item type:', {
                    keys: Object.keys(item),
                    hasAssignmentId: !!item.assignmentId,
                    hasResponseId: !!item.responseId,
                    hasTitle: !!item.title,
                    type: item.type
                });
                
                return false;
            });
            
            console.log('Found actual assignments after filtering:', assignments.length);
            
            // Sort by creation date (newest first)
            assignments.sort((a, b) => {
                const dateA = new Date(a.createdAt || '');
                const dateB = new Date(b.createdAt || '');
                return dateB.getTime() - dateA.getTime();
            });
            
            // Log assignment titles for debugging
            console.log('Assignment titles:', assignments.map(a => a.title));
            
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: true,
                    assignments: assignments,
                    count: assignments.length
                })
            };
        }
        
    } catch (error) {
        // Log the error for debugging
        console.error('Error retrieving assignments:', error);
        console.error('Event:', JSON.stringify(event));
        console.error('Error stack:', error.stack);
        
        // Return error response
        return {
            statusCode: 500,
            headers: corsHeaders,
            body: JSON.stringify({
                success: false,
                error: 'Internal server error',
                message: error.message,
                details: error.stack
            })
        };
    }
};